File: solutions/0543/1.java
class Solution {
    int diameter;

    public int diameterOfBinaryTree(TreeNode root) {
        diameter = 0;
        traverse(root);
        return diameter;
    }

    // 返回树的深度
    int traverse(TreeNode root) {
        if (root == null) {
            return 0;
        }

        int left = traverse(root.left); // 左子树的深度
        int right = traverse(root.right); // 右子树的深度
        // 直接访问全局变量
        diameter = Math.max(diameter, left + right);
        return 1 + Math.max(left, right);
    }
}

File: solutions/0111/1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int minDepth(TreeNode* root) {
        if (root == nullptr) {
            return 0;
        }
        if (root->left == nullptr && root->right == nullptr) {
            return 1;
        } else if (root->left == nullptr) {
            return 1 + minDepth(root->right);
        } else if (root->right == nullptr) {
            return 1 + minDepth(root->left);
        } else {
            return 1 + min(minDepth(root->left), minDepth(root->right));
        }
    }
};
